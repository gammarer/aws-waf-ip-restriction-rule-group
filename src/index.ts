import * as waf from 'aws-cdk-lib/aws-wafv2';
import { Construct } from 'constructs';

export interface WafIpRestrictRuleGroupProps {
  readonly name?: string;
  readonly scope: Scope;
  readonly allowIpSetArn: string;
  //readonly rateLimitCount?: number;
  //whitelist
}

export enum Scope {
  GLOBAL = 'Global',
  REGIONAL = 'Regional',
}

export class WafIpRestrictRuleGroup extends waf.CfnRuleGroup {
  constructor(scope: Construct, id: string, props: WafIpRestrictRuleGroupProps) {
    super(scope, id, {
      name: props.name,
      description: 'ip restrict rule group',
      scope: ((): string => {
        switch (props.scope) {
          case Scope.GLOBAL:
            return 'CLOUDFRONT';
          case Scope.REGIONAL:
            return 'REGIONAL';
        }
      })(),
      capacity: 10,
      rules: [
        {
          priority: 0,
          name: 'allow-ip-rule',
          action: {
            allow: {},
          },
          visibilityConfig: {
            cloudWatchMetricsEnabled: true,
            sampledRequestsEnabled: true,
            metricName: 'AllowIpRule',
          },
          statement: {
            ipSetReferenceStatement: {
              arn: props.allowIpSetArn,
            },
          },
        },
        {
          priority: 1,
          name: 'deny-ip-rule',
          action: {
            block: {
              CustomResponse: {
                CustomResponseBodyKey: 'ip-restrict',
                ResponseCode: 403,
              },
            },
          },
          visibilityConfig: {
            cloudWatchMetricsEnabled: true,
            sampledRequestsEnabled: true,
            metricName: 'DenyIpRule',
          },
          statement: {
            notStatement: {
              statement: {
                ipSetReferenceStatement: {
                  arn: props.allowIpSetArn,
                },
              },
            },
          },
        },
      ],
      customResponseBodies: {
        ['ip-restrict']: {
          contentType: 'TEXT_PLAIN',
          content: 'Sorry, You Are Not Allowed to Access This Service.',
        },
      },
      visibilityConfig: {
        cloudWatchMetricsEnabled: true,
        sampledRequestsEnabled: true,
        metricName: 'IpRestrictRule',
      },
    });
  }
}