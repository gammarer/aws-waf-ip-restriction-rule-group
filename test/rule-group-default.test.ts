import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import { Scope, WafIpRestrictRuleGroup } from '../src';

describe('Web ACL Rule Group default testing', () => {

  describe('Web Acl Rule Group scope=global testing', () => {

    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const allowIpSet = new waf.CfnIPSet(stack, 'IpSet', {
      addresses: [
        '192.0.2.0/24',
        '198.51.100.0/24',
        '203.0.113.0/24',
      ],
      ipAddressVersion: 'IPV4',
      scope: 'CLOUDFRONT',
    });

    const ruleGroup = new WafIpRestrictRuleGroup(stack, 'WafIpRestrictRuleGroup', {
      scope: Scope.GLOBAL,
      allowIpSetArn: allowIpSet.attrArn,
    });

    it('Is Waf RuleGroup', async () => {
      expect(ruleGroup).toBeInstanceOf(waf.CfnRuleGroup);
    });

    const template = Template.fromStack(stack);

    it('Should have WAF Rule Group', async () => {
      template.hasResourceProperties('AWS::WAFv2::RuleGroup', Match.objectEquals({
        Description: 'ip restrict rule group',
        Scope: 'CLOUDFRONT',
        Capacity: 10,
        CustomResponseBodies: {
          'ip-restrict': {
            Content: 'Sorry, You Are Not Allowed to Access This Service.',
            ContentType: 'TEXT_PLAIN',
          },
        },
        Rules: Match.arrayWith([
          {
            Priority: 0,
            Name: 'allow-ip-rule',
            Action: {
              Allow: {},
            },
            VisibilityConfig: {
              CloudWatchMetricsEnabled: true,
              MetricName: 'AllowIpRule',
              SampledRequestsEnabled: true,
            },
            Statement: {
              IPSetReferenceStatement: {
                Arn: {
                  'Fn::GetAtt': ['IpSet', 'Arn'],
                },
              },
            },
          },
          {
            Priority: 1,
            Name: 'deny-ip-rule',
            Action: {
              Block: {
                CustomResponse: {
                  CustomResponseBodyKey: 'ip-restrict',
                  ResponseCode: 403,
                },
              },
            },
            VisibilityConfig: {
              CloudWatchMetricsEnabled: true,
              MetricName: 'DenyIpRule',
              SampledRequestsEnabled: true,
            },
            Statement: {
              NotStatement: {
                Statement: {
                  IPSetReferenceStatement: {
                    Arn: {
                      'Fn::GetAtt': ['IpSet', 'Arn'],
                    },
                  },
                },
              },
            },
          },
        ]),
        VisibilityConfig: {
          CloudWatchMetricsEnabled: true,
          MetricName: 'IpRestrictRule',
          SampledRequestsEnabled: true,
        },
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('global');
    });

  });

  describe('Web Acl Rule Group scope=regional testing', () => {

    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    const allowIpSet = new waf.CfnIPSet(stack, 'IpSet', {
      addresses: [
        '192.0.2.0/24',
        '198.51.100.0/24',
        '203.0.113.0/24',
      ],
      ipAddressVersion: 'IPV4',
      scope: 'REGIONAL',
    });

    const ruleGroup = new WafIpRestrictRuleGroup(stack, 'WafIpRestrictRuleGroup', {
      name: 'example-waf-ip-restrict-rule-group',
      scope: Scope.REGIONAL,
      allowIpSetArn: allowIpSet.attrArn,
    });

    it('Is Waf RuleGroup', async () => {
      expect(ruleGroup).toBeInstanceOf(waf.CfnRuleGroup);
    });

    const template = Template.fromStack(stack);

    it('Should have WAF Rule Group', () => {
      template.hasResourceProperties('AWS::WAFv2::RuleGroup', Match.objectEquals({
        Name: 'example-waf-ip-restrict-rule-group',
        Description: 'ip restrict rule group',
        Scope: 'REGIONAL',
        Capacity: 10,
        CustomResponseBodies: {
          'ip-restrict': {
            Content: 'Sorry, You Are Not Allowed to Access This Service.',
            ContentType: 'TEXT_PLAIN',
          },
        },
        Rules: Match.arrayWith([
          {
            Priority: 0,
            Name: 'allow-ip-rule',
            Action: {
              Allow: {},
            },
            VisibilityConfig: {
              CloudWatchMetricsEnabled: true,
              MetricName: 'AllowIpRule',
              SampledRequestsEnabled: true,
            },
            Statement: {
              IPSetReferenceStatement: {
                Arn: {
                  'Fn::GetAtt': ['IpSet', 'Arn'],
                },
              },
            },
          },
          {
            Priority: 1,
            Name: 'deny-ip-rule',
            Action: {
              Block: {
                CustomResponse: {
                  CustomResponseBodyKey: 'ip-restrict',
                  ResponseCode: 403,
                },
              },
            },
            VisibilityConfig: {
              CloudWatchMetricsEnabled: true,
              MetricName: 'DenyIpRule',
              SampledRequestsEnabled: true,
            },
            Statement: {
              NotStatement: {
                Statement: {
                  IPSetReferenceStatement: {
                    Arn: {
                      'Fn::GetAtt': ['IpSet', 'Arn'],
                    },
                  },
                },
              },
            },
          },
        ]),
        VisibilityConfig: {
          CloudWatchMetricsEnabled: true,
          MetricName: 'IpRestrictRule',
          SampledRequestsEnabled: true,
        },
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('regional');
    });

  });

});