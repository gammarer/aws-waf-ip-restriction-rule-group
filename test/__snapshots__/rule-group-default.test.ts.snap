// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web ACL Rule Group default testing Web Acl Rule Group scope=global testing Should match snapshot: global 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "IpSet": {
      "Properties": {
        "Addresses": [
          "192.0.2.0/24",
          "198.51.100.0/24",
          "203.0.113.0/24",
        ],
        "IPAddressVersion": "IPV4",
        "Scope": "CLOUDFRONT",
      },
      "Type": "AWS::WAFv2::IPSet",
    },
    "WafIpRestrictRuleGroup": {
      "Properties": {
        "Capacity": 10,
        "CustomResponseBodies": {
          "ip-restrict": {
            "Content": "Sorry, You Are Not Allowed to Access This Service.",
            "ContentType": "TEXT_PLAIN",
          },
        },
        "Description": "ip restrict rule group",
        "Rules": [
          {
            "Action": {
              "Allow": {},
            },
            "Name": "allow-ip-rule",
            "Priority": 0,
            "Statement": {
              "IPSetReferenceStatement": {
                "Arn": {
                  "Fn::GetAtt": [
                    "IpSet",
                    "Arn",
                  ],
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AllowIpRule",
              "SampledRequestsEnabled": true,
            },
          },
          {
            "Action": {
              "Block": {
                "CustomResponse": {
                  "CustomResponseBodyKey": "ip-restrict",
                  "ResponseCode": 403,
                },
              },
            },
            "Name": "deny-ip-rule",
            "Priority": 1,
            "Statement": {
              "NotStatement": {
                "Statement": {
                  "IPSetReferenceStatement": {
                    "Arn": {
                      "Fn::GetAtt": [
                        "IpSet",
                        "Arn",
                      ],
                    },
                  },
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "DenyIpRule",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "CLOUDFRONT",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "IpRestrictRule",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::RuleGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Web ACL Rule Group default testing Web Acl Rule Group scope=regional testing Should match snapshot: regional 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "IpSet": {
      "Properties": {
        "Addresses": [
          "192.0.2.0/24",
          "198.51.100.0/24",
          "203.0.113.0/24",
        ],
        "IPAddressVersion": "IPV4",
        "Scope": "REGIONAL",
      },
      "Type": "AWS::WAFv2::IPSet",
    },
    "WafIpRestrictRuleGroup": {
      "Properties": {
        "Capacity": 10,
        "CustomResponseBodies": {
          "ip-restrict": {
            "Content": "Sorry, You Are Not Allowed to Access This Service.",
            "ContentType": "TEXT_PLAIN",
          },
        },
        "Description": "ip restrict rule group",
        "Name": "example-waf-ip-restrict-rule-group",
        "Rules": [
          {
            "Action": {
              "Allow": {},
            },
            "Name": "allow-ip-rule",
            "Priority": 0,
            "Statement": {
              "IPSetReferenceStatement": {
                "Arn": {
                  "Fn::GetAtt": [
                    "IpSet",
                    "Arn",
                  ],
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "AllowIpRule",
              "SampledRequestsEnabled": true,
            },
          },
          {
            "Action": {
              "Block": {
                "CustomResponse": {
                  "CustomResponseBodyKey": "ip-restrict",
                  "ResponseCode": 403,
                },
              },
            },
            "Name": "deny-ip-rule",
            "Priority": 1,
            "Statement": {
              "NotStatement": {
                "Statement": {
                  "IPSetReferenceStatement": {
                    "Arn": {
                      "Fn::GetAtt": [
                        "IpSet",
                        "Arn",
                      ],
                    },
                  },
                },
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "DenyIpRule",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "REGIONAL",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": true,
          "MetricName": "IpRestrictRule",
          "SampledRequestsEnabled": true,
        },
      },
      "Type": "AWS::WAFv2::RuleGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
